### PR TITLE
Fix flaky tests for deprecation

### PR DESCRIPTION
**Pull request by johscheuer**
Fixes: https://github.com/FoundationDB/fdb-kubernetes-operator/issues/401

@rbtcollins PTAL

### CODE REVIEW

File: controllers/cluster_controller_test.go [hunk 1]
```diff
@@ -3158,7 +3158,9 @@ var _ = Describe("cluster_controller", func() {
 		JustBeforeEach(func() {
 			err := k8sClient.Create(context.TODO(), cluster)
 			Expect(err).NotTo(HaveOccurred())
-			Eventually(func() (int64, error) { return reloadCluster(cluster) }, 1).Should(Equal(int64(1)))
+			Eventually(func() (int64, error) {
+				return reloadCluster(cluster)
+			}, 5*time.Second).Should(Equal(int64(1)))
```
Comment (johscheuer) [inline controllers/cluster_controller_test.go:3163]: The issue actually one happened here but I think the other statements are also fine and should make the test a little bit more robust.

**Comment by rbtcollins** [comment]
So this is giving all the assertions 5 seconds to complete rather than assuming instantaneous completion, yes? Seems good to me.

**Comment by rbtcollins** [comment]
Test runs with this patch applied.

1:
```


Summarizing 1 Failure:

[Fail] cluster_controller Reconciliation change the storage servers per disk [It] should replace the pods 
/home/robertc/src/fdb-kubernetes-operator/controllers/cluster_controller_test.go:259
```

run 2: passed

run 3: passed

**Comment by rbtcollins** [comment]
I have had a look at the timeouts for tests though, and they are extremely hard to reason about - there are many timeout settings that have no effect because they have been made irrelevant by rearranged code, or are overridden by deeper nested context's BeforeEach's. I'm preparing a patch now - 

```
git diff --stat
 controllers/admin_client_test.go       |  4 +---
 controllers/backup_controller_test.go  | 17 +++++++----------
 controllers/cluster_controller_test.go | 78 ++++++++++++++++++++++++++++++++----------------------------------------------
 controllers/restore_controller_test.go |  8 ++------
 controllers/suite_test.go              |  1 +
```

**Comment by johscheuer** [comment]
Should I close my PR if you provide a patch or is it additional?

**Comment by rbtcollins** [comment]
I've layered it on your branch as it happens, I think merging yours makes sense.

I'm pondering perhaps just making everything timeout with 10s - the test suite will slowly get slower because theres no backpressure, but unless folk are reacting to failures by optimising, all we're really doing by having failures is making the creeping slowness get annotated by sprinkled timeout values throughout the code anyway.

**Comment by johscheuer** [comment]
I mean we could also revisit our reconcile loops and try to avoid writing the same object (mostly the cluster) in multiple places.

**Comment by rbtcollins** [comment]
Looking at the checks on this PR - 2m51s to build and test; bot for me just the test run itself is taking 6m36s for me, and CPU is effectively 0: IO is low (100's of K/sec) but jbd2 is pinned at 99.99% wait, and etcd is spending 25% or more time in IO wait states:

<details>
<summary>iotop output</summary>

```
    609 be/3 root        0.00 B/s   21.50 K/s  0.00 % 99.99 % [jbd2/dm-2-8]
  64719 be/4 robertc     0.00 B/s   39.42 K/s  0.00 % 16.16 % etcd --listen-peer-urls=http://localhost:0 --advertise-client-urls=http://127.0.0.1:42219 --listen-client-urls=http://127.0.0.1:42219 --data-dir=/tmp/k8s_test_framework_531258979
  64717 be/4 robertc     0.00 B/s   50.17 K/s  0.00 % 15.15 % etcd --listen-peer-urls=http://localhost:0 --advertise-client-urls=http://127.0.0.1:42219 --listen-client-urls=http://127.0.0.1:42219 --data-dir=/tmp/k8s_test_framework_531258979
  64724 be/4 robertc     0.00 B/s   46.58 K/s  0.00 % 13.33 % etcd --listen-peer-urls=http://localhost:0 --advertise-client-urls=http://127.0.0.1:42219 --listen-client-urls=http://127.0.0.1:42219 --data-dir=/tmp/k8s_test_framework_531258979
  64767 be/4 robertc     0.00 B/s   50.17 K/s  0.00 % 10.58 % etcd --listen-peer-urls=http://localhost:0 --advertise-client-urls=http://127.0.0.1:42219 --listen-client-urls=http://127.0.0.1:42219 --data-dir=/tmp/k8s_test_framework_531258979
  64714 be/4 robertc     0.00 B/s    7.17 K/s  0.00 %  4.75 % etcd --listen-peer-urls=http://localhost:0 --advertise-client-urls=http://127.0.0.1:42219 --listen-client-urls=http://127.0.0.1:42219 --data-dir=/tmp/k8s_test_framework_531258979
  64721 be/4 robertc     0.00 B/s   25.08 K/s  0.00 %  4.24 % etcd --listen-peer-urls=http://localhost:0 --advertise-client-urls=http://127.0.0.1:42219 --listen-client-urls=http://127.0.0.1:42219 --data-dir=/tmp/k8s_test_framework_531258979
```

</details>

The underlying disk for this VM is a HDD, so fsync is relatively slow.

putting eatmydata in the mix increased IO bandwidth for the test suite by 10x - disk writes went up to 1M/sec and percentages dropped.

<details>
<summary>iotop output</summary>

```
Total DISK READ:         0.00 B/s | Total DISK WRITE:       650.17 K/s
Current DISK READ:       0.00 B/s | Current DISK WRITE:    1053.48 K/s
    TID  PRIO  USER     DISK READ  DISK WRITE  SWAPIN     IO>    COMMAND
    609 be/3 root        0.00 B/s   13.91 K/s  0.00 % 43.00 % [jbd2/dm-2-8]
  67502 be/4 robertc     0.00 B/s   38.25 K/s  0.00 % 13.13 % etcd --listen-peer-urls=http://localhost:~ata-dir=/tmp/k8s_test_framework_816236212
  67445 be/4 robertc     0.00 B/s  118.21 K/s  0.00 % 10.77 % etcd --listen-peer-urls=http://localhost:~ata-dir=/tmp/k8s_test_framework_816236212
  67449 be/4 robertc     0.00 B/s  111.26 K/s  0.00 %  9.42 % etcd --listen-peer-urls=http://localhost:~ata-dir=/tmp/k8s_test_framework_816236212
  67570 be/4 robertc     0.00 B/s  139.07 K/s  0.00 %  9.32 % etcd --listen-peer-urls=http://localhost:~ata-dir=/tmp/k8s_test_framework_816236212
  67455 be/4 robertc     0.00 B/s   48.68 K/s  0.00 %  5.85 % etcd --listen-peer-urls=http://localhost:~ata-dir=/tmp/k8s_test_framework_816236212
  68624 be/4 robertc     0.00 B/s  132.12 K/s  0.00 %  3.94 % etcd --listen-peer-urls=http://localhost:~ata-dir=/tmp/k8s_test_framework_816236212
  68578 be/4 robertc     0.00 B/s   20.86 K/s  0.00 %  1.35 % etcd --listen-peer-urls=http://localhost:~ata-dir=/tmp/k8s_test_framework_816236212
  67528 be/4 robertc     0.00 B/s   27.81 K/s  0.00 %  1.14 % etcd --listen-peer-urls=http://localhost:~ata-dir=/tmp/k8s_test_framework_816236212
```

</details>
Wall clock time for the test run was not really affected.

Next I tried a ramdisk - mounting tmpfs and redirecting TMPDIR there.

This worked much better:
```
real    1m18.311s
user    1m55.278s
sys     0m47.055s
```

though CPU use is still horribly low - only about 1.7 cores used at peak during the test run, so at least in principle this could be a 3-4 second test execution if it was parallelisable.

However, `time ginko -p controllers` currently fails due to the use of a fixed port address for the stub servers: 
```

  Unexpected error:
      <*errors.errorString | 0xc00085bd80>: {
          s: "error listening on :8080: listen tcp :8080: bind: address already in use",
      }
      error listening on :8080: listen tcp :8080: bind: address already in use
  occurred

  /home/robertc/src/fdb-kubernetes-operator/controllers/suite_test.go:88
```

so thats going to have to wait for another day.

I'll push my patch up tomorrow (its 1am) and include some dev docs for making this fast on machines which don't have lots of IO available by default.

My guess is that the flakes in CI have been due to IO contention with noisy neighbours.

**Comment by johscheuer** [comment]
Thanks @rbtcollins for your analyse! I really appreciate it and I know that there are places where we can improve the test suite. Would you mind to fill out issues for everything you find and we can work on fixing them?

**Comment by rbtcollins** [comment]
I've put up my stuff so far in https://github.com/FoundationDB/fdb-kubernetes-operator/pull/407

**Comment by johscheuer** [comment]
Closed in favour of: https://github.com/FoundationDB/fdb-kubernetes-operator/pull/407

**Pull request closed without merging**